vsg::Object cast<T>() sometimes returns nullptr on macOS Sonoma arm64 Release builds

hearga opened the pull request:
## Description

- Addressed an issue where vsg::Object::cast<T>() would sometimes return nullptr on macOS Sonoma (Apple Silicon, arm64) in Release builds.
- Updated the casting logic to use dynamic_cast<T*>(), matching the approach already used  in the previous function " template<class T>
      ref_ptr<T> readObject(const char* propertyName)"

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)

robertosfield commented:
We need to get to the bottom of why cast<> is failing, not go around relacing all of it's usage across the VSG.

Is that are particular object type that is failing?

robertosfield commented:
Could you have a look at this thread, I believe it touches upon the issue you are seeing:

https://github.com/vsg-dev/VulkanSceneGraph/issues/899

It looks like an issue with clang generating different type_info results for different libraries in certain circumstances.  At the end of the above thread there are suggestions for solutions.

hearga commented:
It works using explicit template instantiation. It resolves the different type_info hash IDs during casting.

robertosfield commented:
Thanks for the changes, I've done a first past review and it feels like a rather heavy handed approach, forcing lots of extra code that we don't need other than for working around a problem with a specific compiler's build settings.

The changes would also need to be rolled out to all the other use of templates such as vsg::Value, Array2D, Array3D and perhaps even use of vsg::Inherit.

Did you test building as dynamic library?

hearga commented:
Thanks for the review, I’ve reworked this to a minimal fix that targets the actual Clang issue (hidden visibility causing divergent &typeid(Array<T>) across TUs):

- Export the class: Array<T> to stabilize typeid.
- No explicit instantiations: removed ArrayInstantiations.cpp.
- Opt-in escape hatch: VSG_USE_DYNAMIC_CAST switches cast<T>() to dynamic_cast for those who prefer it.

If this approach looks good, we can apply the same small change to Value, Array2D/Array3D, etc., as needed, no project-wide mass changes required.

Shared libs: not tested, sorry, I have limited time to work on this; in shared builds, we already define the import/export macro, so the type info should be default-visible and typeid stable—no additional changes expected.

robertosfield commented on src/vsg/CMakeLists.txt:
```diff
@@ -426,6 +426,14 @@ target_link_libraries(vsg ${LIBRARIES})
     target_compile_definitions(vsg INTERFACE VSG_SHARED_LIBRARY)
+else()
+    if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+        if (VSG_USE_DYNAMIC_CAST)
```
Where in CMake scripts is VSG_USE_DYNAMIC_CAST set & controlled?  Is there a missing CMake option for this?

robertosfield commented on include/vsg/core/Array.h:
```diff
@@ -33,7 +33,7 @@ namespace vsg
 
     template<typename T>
-    class Array : public Data
+    class VSG_DECLSPEC Array : public Data
```
If Array requires  VSG_DECLSPEC then Value, Array2D and Array3D will also require it.

robertosfield commented:
After a crazy busy period of client work I'm now back catching up with PRs, sorry for taking so long to get back to this.

I've done another code review and added comments into the various changes.  I think some of the changes might be sensible to merge separately so if there are ones that aren't directly dependent on the dynamic_cast<> usage then it would be good to get these fixed as separate commits/PRs.  I'm happy to do some of this, just point me in the right direction.

robertosfield commented:
After review the CMake and Version.h.in I have changed to use CMake option and changed the capitalization of the define to be a more direct about what it maps too: 0d7cc5ea8d9fa3df4b155a58b2b4e5fd0230a63e
